AIP-86 - Add doc pages for the new Deadline Alerts feature

ferruzzi opened the pull request:
With the Deadline Alerts reaching minimum-launch requirements (hopefully) next week, I want to get the docs reviewed and finalized.  The more eyes, the better,  Thanks!  I'll also be following this up with a more plain-English blog/announcement once we hash out any phrasing issues we may find here.  Below you will see the rendered output from build-docs, which may be easier to review.  I'll try to keep those screenshots updated as I make adjustments, but of course the code is the source of truth if there is a mismatch, not the screenshots.

I've also included the "significant" newsfragment in this PR for the feature which I'll rename once I submit this and have a pr number. [EDIT:  file renamed]


Screenshots below are accurate as of commit [75bf130](https://github.com/apache/airflow/pull/53727/commits/75bf130768048086c20d5e3b1e78e066a56c3a22)

The rendered addition to the dags.rst file:
<img width="2017" height="1041" alt="image" src="https://github.com/user-attachments/assets/d27814ba-e21b-458a-9a1f-385ba46b2bc6" />

The changed section of tasks.rst:
<img width="1157" height="223" alt="image" src="https://github.com/user-attachments/assets/2a6fe7cc-2bb7-4c88-b661-b3c1cdf2897f" />


The rendered new howto/deadline-alerts.rst file:
<img width="2560" height="8494" alt="image" src="https://github.com/user-attachments/assets/d06c044e-d3f1-4dd8-8fc5-0522f71d7937" />

ferruzzi committed:
AIP-86 - Add doc pages for the new Deadline Alerts feature

ferruzzi committed:
Rename the news fragment now that I have a PR number to reference

ferruzzi committed:
Static checks and changed the news fragment to "feature" type.

ferruzzi committed:
Update the docs based on end-to-end testing and community discussions.

Notably:
- Jinja templates don't actually work (yet) based on manual testing
- Clarified that the callbacks must be available to the Triggerer, and based on community discussion, the dag folder is not the right place for this.
- Updated the code snippet to indicate these points

ferruzzi committed:
Static checks in a doc-only change.  TIL  Makes sense

vincbeck commented on airflow-core/docs/core-concepts/dags.rst:
```diff
@@ -829,3 +829,40 @@ if it fails for ``N`` number of times consecutively.
+calculated references (like DAG queue time or start time), or implement your own custom reference.
+When a deadline is exceeded, it triggers a callback which can notify you or take other actions.
+
+Here's a simple example using the existing email Notifier:
```
Unless you have good reason to keep it capitalized?

```suggestion
Here's a simple example using the existing email notifier:
```

ferruzzi replied:
Do we not use the proper noun there?   I'll see if I can find precedent in other doc pages, but I think we usually capitalize things lie Operator, Sensor, Notifier, etc.  I'll double check.

o-nikolas replied:
I like capitalizing the proper noun here, FWIW

vincbeck commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,255 @@
+===============
+
+The :class:`~airflow.sdk.definitions.deadline.DeadlineAlert` feature is the next evolution of
+the old SLA.  Deadline Alerts allow you to set time thresholds for your DAG runs and automatically
```
I think it has been brought up we should not compare Airflow version in docs, I am just wondering if mentioning "the old SLA" goes against that.

shahar1 replied:
Currently, the [Tasks docs](https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/tasks.html#slas) metion:

```The SLA feature from Airflow 2 has been removed in 3.0 and will be replaced with a new implementation in Airflow 3.1```

So we could update the above to something like 

```The SLA feature from Airflow 2 has been removed in 3.0 and was replaced in Airflow 3.1 with Deadlines Alerts (+link)```

And delete the comparison in this section (```The :class:`~airflow.sdk.definitions.deadline.DeadlineAlert` feature is the next evolution of
the old SLA.```).

What do you think?

ferruzzi replied:
I think it's a good solution.  I'll make it happen.  If someone has a better solution, we can always change it again later.

vincbeck commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,255 @@
+
+    |------|-----------|---------|-----------|--------|
+        Scheduled    Queued    Started    Deadline
+         00:00       00:03      00:05      00:18
```
It is 15 minutes right? Other than that, I love this section, it makes it super easy to understand the concept of deadlines!

```suggestion
         00:00       00:03      00:05      00:15
```

ferruzzi replied:
Yup, but it's 15 minutes after the QUEUED_AT in this example, so 0:18 is right.  If you used the LOGICAL_DATE reference i would be at 0:15

vincbeck replied:
Oh! My bad, I was not attentive enough

vincbeck commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,255 @@
+        dag_id="fixed_deadline_alert",
+        deadline=DeadlineAlert(
+            reference=DeadlineReference.FIXED_DATETIME(tomorrow_at_ten),
+            interval=timedelta(minutes=-30),  # Alert 30 minutes before the deadline
```
```suggestion
            interval=timedelta(minutes=-30),  # Alert 30 minutes before the reference
```

ferruzzi replied:
That's fair, I'll make the change.

vincbeck commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,255 @@
+
+    |------|----------|---------|----------|--------|
+        Deadline    Queued    Start    Reference
+         08:00      09:15     09:17     10:00
```
Is this timeline right? It does not seem to match the description above

ferruzzi replied:
Oh man, you are right, I changed the example and never updated the timeline!  Good catch, thanks!  It should be 

```
    |------|------------|----------|--------------|--------|
        Queued    Start   Deadline   Reference
          09:15      09:17       09:30       10:00
```         

Given that this one is "backwards" (negative interval)  and the deadline is a time before the reference, do you think it is any easier to understand with an arrow or some indicator like this:

```
    |------|------------|----------|-----------------|--------|
        Queued    Start    Deadline       Reference
          09:15      09:17       09:30   <==   10:00
```

vincbeck replied:
I understood without it so I am not sure this is needed, but it does not harm having it either. I'll let you decide :)

ferruzzi replied:
Alright, for mow I'll just fix the layout and leave off the arrow.  Thanks.

vincbeck commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,255 @@
+        dag_id="slack_deadline_alert",
+        deadline=DeadlineAlert(
+            reference=DeadlineReference.DAGRUN_QUEUED_AT,
+            interval=timedelta(minutes=15),
```
```suggestion
            interval=timedelta(minutes=30),
```

ferruzzi replied:
Thanks so much for catching these.  I went through a bunch of revisions and clearly missed some edits.  A bit embarrassing....

vincbeck replied:
No worries :) It is always harder to catch these when you're head down :)

vincbeck reviewed (commented):
Some minor comments. But overall I really like the doc, by reading it, it allowed me to understand better the topic of deadlines

ferruzzi committed:
Implement suggestions from Vincent and Ramit

Dev-iL commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,263 @@
+Creating a Deadline Alert
+-------------------------
+
+To create a Deadline Alert, there are three three components you must specify and one optional one:
```
```suggestion
To create a Deadline Alert, there are three components you must specify and one optional one:
```

Dev-iL replied:
Alternative phrasing: "three mandatory and one optional parameter"

ferruzzi replied:
Thanks, I'll get it in the upcoming revision.  I've changed it to "three required...." which I think is what my brain meant to type but my fingers didn't get the memo.

o-nikolas commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,263 @@
+Using Callbacks
+---------------
+
+When a deadline is exceeded, the callback is executed. You can use any async :doc:`Notifier </howto/notifications>`
```
```suggestion
When a deadline is exceeded, the callback is executed. You can use any :doc:`Notifier </howto/notifications>`
```
Using sync or async is the goal right? Did you specifically mention async here because the sync path is still in progress? If so I'd just do something like:


```suggestion
When a deadline is exceeded, the callback is executed. You can use any :doc:`Notifier </howto/notifications>` (Note: that synchronous callback support is in development)
```

ferruzzi replied:
I actually have that adjusted in a current revision.  Yes, that was my reasoning, but we're going to fast-track the sync work so I am tweaking it now.

ferruzzi replied:
What I have there right now is 

===

When a deadline is exceeded, the callback is executed. You can use an existing :doc:`Notifier </howto/notifications>` or create a custom callback function.

===

I'm just waiting for build-docs and static checks to run and I'll be pushing that change and a rework of the custom callbacks section to add more non-async context.

o-nikolas commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,263 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can create custom callbacks for more complex handling. The ``callback_kwargs`` specified in
+the ``DeadlineAlert`` are passed to the callback function.
```
This and the code snippets need updating after https://github.com/apache/airflow/pull/53951#pullrequestreview-3080641355 right?

ferruzzi replied:
That is possible.   I'll keep an eye on the in-flight PRs and adjust accordingly.

ferruzzi replied:
Updated in https://github.com/apache/airflow/pull/53727/commits/28d88e269bbb2123557f75f4716550c95981eb8a

o-nikolas commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,263 @@
+    * The Triggerer may need to be restarted when a callback is added or changed in order to reload the files.
+
+
+Deadline Calculation
```
This block is really nice to hit home the notion of how this all works. Maybe even pull it closer to the top.

ferruzzi committed:
Implement suggestions from Niko and Iliya

Also removed reference to Jinja templating as that does not look like it will be ready for launch.

ferruzzi committed:
Update docs for BaseCallback/AsyncCallback/SyncCallback and more accurate examples

gopidesupavan commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+            reference=DeadlineReference.FIXED_DATETIME(tomorrow_at_ten),
+            interval=timedelta(minutes=-30),  # Alert 30 minutes before the reference.
+            callback=AsyncCallback(
+                SlackWebhookNotifier,
```
This should be awaitable for `AsyncCallback ` , SlackWebhookNotifier doesn't have any awaitbale's. IMHO its better to keep working example ? WDYT?

ramitkataria replied:
Right now in `main`, there's no awaitable Notifiers but once #53831 is merged `SlackWebhookNotifier` will be awaitable and we'll make a PR for a couple more awaitable notifiers as well. So I think the plan is to merge that PR first (some unit tests still require some work) and then merge this one so it will be a working example

ferruzzi replied:
Hey, sorry, I've been on a camping trip.   First, thanks for the review and I appreciate the catch!  Second, what Ramit said :P   At this point, this is kind of a "living document" since it's documenting a feature that is almost but not quite done.  See the comments below about callback_kwargs as another example.  Tha Notifier will be awaitable before this is merged though.

gopidesupavan commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+            reference=DeadlineReference.DAGRUN_QUEUED_AT,
+            interval=timedelta(minutes=30),
+            callback=AsyncCallback(
+                SlackNotifier,
```
Same as above comment.

ramitkataria replied:
I think we should use `SlackWebhookNotifier` here as well instead of `SlackNotifier` because that's the one we're merging first and will be a working example when this PR is merged

ferruzzi replied:
Going to unresvolve this one for a bit so i remember to make Ramit's change.  Thanks for catching it!

hussein-awala reviewed (approved):
Great work!

Lee-W commented on airflow-core/docs/core-concepts/dags.rst:
```diff
@@ -829,3 +829,40 @@ if it fails for ``N`` number of times consecutively.
+
+.. versionadded:: 3.1
+
+Deadline Alerts allow you to set time thresholds for your DAG runs and automatically respond when those
```
```suggestion
Deadline Alerts allow you to set time thresholds for your Dag runs and automatically respond when those
```

ferruzzi replied:
I'm going to leave one of these open and resolve the rest; this is a simple find/replace.  Thanks for the comment/review.  I'm just getting back from a camping trip but I'll try to get a revision up today or tomorrow with these fixes.

Lee-W replied:
Sounds good 🙂 These can be all replaced as Dag before we merge the PR

Lee-W replied:
I noticed other files have been updated. This one might be missed

ferruzzi replied:
[That file](https://github.com/apache/airflow/blob/main/airflow-core/docs/core-concepts/dags.rst) uses DAG all over the place right now, maybe it's kind of a mess of DAG vs Dag.  I'll fix my additions in this PR and maybe I can do another one quick for the other instances later.

Lee-W commented on airflow-core/docs/core-concepts/dags.rst:
```diff
@@ -829,3 +829,40 @@ if it fails for ``N`` number of times consecutively.
+
+Deadline Alerts allow you to set time thresholds for your DAG runs and automatically respond when those
+thresholds are exceeded. You can set deadlines relative to a fixed datetime, use one of the available
+calculated references (like DAG queue time or start time), or implement your own custom reference.
```
```suggestion
calculated references (like Dag queue time or start time), or implement your own custom reference.
```

Lee-W commented on airflow-core/docs/core-concepts/dags.rst:
```diff
@@ -829,3 +829,40 @@ if it fails for ``N`` number of times consecutively.
+            interval=timedelta(minutes=30),
+            callback=SmtpNotifier(
+                to="team@example.com",
+                subject="[Alert] DAG {{ dag.dag_id }} exceeded time threshold",
```
```suggestion
                subject="[Alert] Dag {{ dag.dag_id }} exceeded time threshold",
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+===============
+
+The :class:`~airflow.sdk.definitions.deadline.DeadlineAlert` feature is the next evolution of
+the old SLA.  Deadline Alerts allow you to set time thresholds for your DAG runs and automatically
```
```suggestion
the old SLA.  Deadline Alerts allow you to set time thresholds for your Dag runs and automatically
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+Creating a Deadline Alert
+-------------------------
+
+Creating a Deadline Alert requires three mandatory and one optional parameter:
```
```suggestion
Creating a Deadline Alert requires 3 mandatory and 1 optional parameter:
```

I feel it might be more readable this way 🤔

ferruzzi replied:
This line is no longer really relevant since https://github.com/apache/airflow/pull/53951/files removed the optional one, I'll be rephrasing it.

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+* Reference: When to start counting from
+* Interval: How far before or after the reference point to trigger the alert
+* Callback: What to do when the deadline is exceeded
+* Callback Kwargs:  Optional values to pass to the Callback when it is run
```
```suggestion
* Callback Kwargs: Optional values to pass to the Callback when it is run
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+* Reference: When to start counting from
+* Interval: How far before or after the reference point to trigger the alert
+* Callback: What to do when the deadline is exceeded
+* Callback Kwargs:  Optional values to pass to the Callback when it is run
```
```suggestion
* ``reference``: When to start counting from
* ``interval``: How far before or after the reference point to trigger the alert
* ``callback``: What to do when the deadline is exceeded
```

looks more like 3 mandatory args but we can modify callback

ferruzzi replied:
Good catch, it needs to be updated; there are now three mandatory and no optional after @ramitkataria 's recent PR merged.

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+        |                                     |
+     Start time                          Trigger point
+
+Below is an example DAG implementation. If the DAG has not finished 15 minutes after it was queued, send a Slack message:
```
```suggestion
Below is an example Dag implementation. If the Dag has not finished 15 minutes after it was queued, send a Slack message:
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+Using Built-in Notifiers
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+Here's an example using the Slack Notifier if the DagRun has not finished within 30 minutes of it being queued:
```
```suggestion
Here's an example using the Slack Notifier if the Dag run has not finished within 30 minutes of it being queued:
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+Creating Custom Callbacks
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can create custom callbacks for more complex handling. The ``callback_kwargs`` specified in
```
This `callback_kwargs` is a bit confusing to me as it's not the actual parameter name.

ferruzzi replied:
It was at the time :P   https://github.com/apache/airflow/pull/53951 removed it, I need to update these docs to match now.  Thanks for pointing it out.

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+        EmptyOperator(task_id="example_task")
+
+A **custom asynchronous callback** is only slightly more work.  Note in the following example that
+the custom callback code is placed in a separate file, and must be imported in the DAG file.
```
```suggestion
the custom callback code is placed in a separate file, and must be imported in the Dag file.
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+        # Additional custom handling here
+
+
+    # Place this in a dag file
```
```suggestion

.. code-block:: python

    # Place this in a dag file
```

maybe making them 2 blocks?

ferruzzi replied:
Sure, I'll make the change in the next revision, thanks.

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+
+.. code-block:: python
+
+    next_meeting = datetime(2025, 06, 26, 9, 30)
```
```suggestion
    next_meeting = datetime(2025, 6, 26, 9, 30)
```

this is a syntax error

ferruzzi replied:
Nice catch, thanks.  I saw that pre-commits ran mypy on the code snippets and had assumed it would have caught that.  Cheers!

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+This will trigger the alert 2 hours before the next meeting starts.
+
+For ``DAGRUN_LOGICAL_DATE``, the interval is typically positive, setting a deadline relative
+to when the DAG was scheduled to run. Here's an example:
```
```suggestion
to when the Dag was scheduled to run. Here's an example:
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+        callback=notify_team,
+    )
+
+In this case, if a DAG is scheduled to run daily at midnight, the deadline would be triggered
```
```suggestion
In this case, if a Dag is scheduled to run daily at midnight, the deadline would be triggered
```

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+    )
+
+In this case, if a DAG is scheduled to run daily at midnight, the deadline would be triggered
+if the DAG hasn't completed by 1:00 AM. This is useful for ensuring that scheduled jobs complete
```
```suggestion
if the Dag hasn't completed by 1:00 AM. This is useful for ensuring that scheduled jobs complete
```

ferruzzi committed:
Revisions for Lee-W and PR 53951

ferruzzi committed:
Revisions for Vincent and Shahar

ferruzzi committed:
fix typo

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+^^^^^^^^^^^^^^^^^^^^^^^^^
+
+You can create custom callbacks for more complex handling. The ``kwargs`` specified in the ``Callback``
+are passed to the callback function, if any are provided.  **Synchronous callbacks** (standard python
```
```suggestion
are passed to the callback function, if any are provided. **Synchronous callbacks** (standard python
```

ferruzzi replied:
Hey, stop picking on my age!  Double-spacer for life!  ((I'll change them))

Lee-W commented on airflow-core/docs/howto/deadline-alerts.rst:
```diff
@@ -0,0 +1,312 @@
+
+You can create custom callbacks for more complex handling. The ``kwargs`` specified in the ``Callback``
+are passed to the callback function, if any are provided.  **Synchronous callbacks** (standard python
+methods) can be defined in the dag bundle and are run in the Executor.  **Asynchronous callbacks** must
```
```suggestion
methods) can be defined in the dag bundle and are run in the Executor. **Asynchronous callbacks** must
```

Lee-W reviewed (approved):
rest are nitpicks. Thanks @ferruzzi !

ferruzzi committed:
Fixes for Lee-W
